### PR TITLE
Fix required components documentation ambiguity

### LINKED ISSUE #16494: Misleading docs on duplicate required component constructors
The [required components documentation](https://dev-docs.bevyengine.org/bevy/ecs/component/trait.Component.html#required-components) describes the algorithm for resolving multiple constructors for the same component in a required components tree as follows:

> In general, this shouldn’t happen often, but when it does the algorithm is simple and predictable:
>
> 1. Use all of the constructors (including default constructors) directly defined in the spawned component’s require list
> 2. In the order the requires are defined in #[require()], recursively visit the require list of each of the components in the list (this is a Depth First Search). When a constructor is found, it will only be used if one has not already been found.

However, this is not what actually happens. For example, consider this case:

```rust
#[derive(Component)]
#[require(B)]
struct A;

#[derive(Component, Default)]
#[require(C, D(|| D(1)))]
struct B;

#[derive(Component, Default)]
#[require(D(|| D(2)))]
struct C;

#[derive(Component)]
struct D(u8);
```

Now suppose we spawn `A`. Following the steps described in the docs to determine which constructor will be used for `D`:

1. The "spawned component" is `A`, which does not have a constructor for `D`, so go to step 2.
2. A DFS will visit the components in the order `A B C D(2) D(1)`. So when `D(2)` is visited, it will be used because `D` hasn't been found yet; and when `D(1)` is visited, it will not be used because `D` has been found already.

This suggests that `D(2)` will be used as the constructor. But that's incorrect: `D(1)` is actually used.

### PR DESCRIPTION
**Pull request by IllyStable**
# Objective

- Fixes #16494

## Solution

- Changed depth first to breadth first (forgive me if I'm wrong, this is my first PR here and I'm fairly new to the Bevy codebase)

## Testing

- I don't believe testing was necessary - one word change

### CODE REVIEW

**Comment by IllyStable** [comment]
Sorry about the one word change, didn't really know how to rephrase it to represent how required components actually propogate

**Review by BenjaminBrienen — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
I don't think that this is quite the right phrasing; it's quite subtle. I'll try and take a crack at this and make a suggestion here, then you tell me if it makes sense to you? :)

File: crates/bevy_ecs/src/component.rs [hunk 1]
```diff
@@ -228,7 +228,7 @@ use derive_more::derive::{Display, Error};
 ///
 /// In general, this shouldn't happen often, but when it does the algorithm is simple and predictable:
 /// 1. Use all of the constructors (including default constructors) directly defined in the spawned component's require list
-/// 2. In the order the requires are defined in `#[require()]`, recursively visit the require list of each of the components in the list (this is a Depth First Search). When a constructor is found, it will only be used if one has not already been found.
+/// 2. In the order the requires are defined in `#[require()]`, recursively visit the require list of each of the components in the list (this is a Breadth First Search). When a constructor is found, it will only be used if one has not already been found.
```
[the remark this answers is not in the document]
Comment (BenjaminBrienen) [reply crates/bevy_ecs/src/component.rs:231]: That sounds a bit off. Maybe:

> ...a direct `#[require()]` are...

[the remark this answers is not in the document]
Comment (BenjaminBrienen) [reply crates/bevy_ecs/src/component.rs:231]: That sounds a bit off. Maybe:

> /// The algorithm for choosing the constructor from the tree:
> /// 1. A constructor from a direct #[requires()], if one exists, is selected with priority.
> /// 2. Otherwise, perform a Depth First Search on the tree of requirements and select the first one found.

[the remark this answers is not in the document]
Comment (benfrankel) [reply crates/bevy_ecs/src/component.rs:231]: I think the main issue with the current wording is that step 1 refers to the "spawned component", which seems to suggest that the step only applies to the component you directly spawn (aka the root of the tree), when it actually applies to every component visited during the DFS. I'd reword it to something roughly like this:

> 1. Starting with the spawned component, recusively visit the require list of each of the components in the list, in the order the requires are defined in `#[require()]` (this is a Depth First Search).
> 2. Whenever a component is visited, use all of the constructors (including default constructors) directly defined in its require list, except for the components that have already been found.

[the remark this answers is not in the document]
Comment (IllyStable) [reply crates/bevy_ecs/src/component.rs:231]: On my readthrough of it I thought "depth-first" was somewhat misleading, but maybe I'm just misunderstanding the term.

[the remark this answers is not in the document]
Comment (benfrankel) [reply crates/bevy_ecs/src/component.rs:231]: The implementation is actually a depth-first search, it just collects the constructors whenever it visits a component.
